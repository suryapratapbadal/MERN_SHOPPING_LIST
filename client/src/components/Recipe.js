import React, { Component } from 'react';
import {
    Collapse,
    Button,
    ListGroup,
    ListGroupItem,
    Progress,
    Card,
    CardBody,
    CardTitle
} from 'reactstrap';

import { CSSTransition, TransitionGroup } from 'react-transition-group';

import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { getRecipes, deleteRecipe } from '../actions/index';


class Recipe extends Component {
    constructor(props) {
        super(props);
        // console.log('....RECIPES PROPS..', this.props)
        this.state = {
            recipes: this.props.recipes,
            status: 'Closed',
            loader: false,
        };
    }

    componentDidUpdate(prevProps) {
        if (this.props.recipes !== prevProps.recipes)
            this.setState({ recipes: this.props.recipes, loader: false })
    }

    // componentWillReceiveProps(nextProps) {
    //     // console.log('.......Collapse....', nextProps, this.props)
    //     this.setState({ recipes: nextProps.recipes })
    // }

    onEntering = () => {
        this.setState({ loader: true })
        this.props.getRecipes(this.props.id)
    }

    // onEntered = () => {
    //     console.log('......Opened....');
    // }

    // onExiting = () => {
    //     console.log('......Closing....');
    // this.setState({recipes: []})
    // }

    // onExited = () => {

    // }

    //   toggle = () => {
    //     this.setState({ collapse: !this.state.collapse });
    //   }

    onDeleteClick = id => {
        this.props.deleteRecipe(id);
    }

    render() {
        const { recipes } = this.state;
        return (
            <div>
                {
                    this.state.loader ?
                        <Progress style={{ marginTop: '2rem', marginBottom: '1rem' }} animated color="info" value={100} /> :
                        <Collapse
                            isOpen={this.props.id === this.props.open_id && this.props.collapse}
                            onEntering={this.onEntering}
                            onEntered={this.onEntered}
                            onExiting={this.onExiting}
                            onExited={this.onExited}
                            style={{ marginTop: '1rem' }}
                        >

                            {recipes.length <= 0 ?
                                <Card>
                                    <TransitionGroup className="recipe-list">
                                        <CSSTransition timeout={500} classNames="fade">
                                            <CardBody>
                                                <CardTitle>No Recipe listed!!</CardTitle>
                                            </CardBody>
                                        </CSSTransition>
                                    </TransitionGroup>
                                </Card> :

                                <ListGroup>
                                    <TransitionGroup className="recipe-list">
                                        {recipes.map(({ _id, name }) =>
                                            (
                                                <CSSTransition key={_id} timeout={500} classNames="fade">
                                                    <ListGroupItem color="info" key={_id}>
                                                        {name}
                                                        <Button style={{ float: 'right' }} className="remove-btn"
                                                            color='danger'
                                                            size="sm"
                                                            onClick={this.onDeleteClick.bind(this, _id)}
                                                        >
                                                            &times;
                                                        </Button>
                                                    </ListGroupItem>
                                                </CSSTransition>
                                            )
                                        )}
                                    </TransitionGroup>
                                </ListGroup>}

                        </Collapse>
                }
            </div>
        );
    }
}

export default connect(state => {
    const recipes = state.recipeReducer.recipes || [];
    return {
        recipes
    }
}, dispatch => {
    return bindActionCreators({ getRecipes: getRecipes, deleteRecipe: deleteRecipe }, dispatch)
}
)(Recipe);