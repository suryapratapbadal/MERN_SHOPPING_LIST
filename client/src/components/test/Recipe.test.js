import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

// Component to be tested
import { Recipe } from '../Recipe';

const mockgetRecipessfn = jest.fn();
const mockdeleteRecipefn = jest.fn();
const mockonEnteringfn = jest.fn();
const recipes = [{ _id: '1', name: 'Tea' }];
const classes = {
    root: {},
    nested: {},
};
const id ='1';

describe('<Recipe />', () => {

    const wrapper = shallow(<Recipe recipes={recipes} classes={classes} deleteRecipe={mockdeleteRecipefn} getRecipes={mockgetRecipessfn} onEntering={mockonEnteringfn} open_id={id} id={id} collapse={true}/>);
    const event = {
        stopPropagation: jest.fn(),
        target: {
            value: 'dummy'
        }
    }
    describe('render()', () => {

        test('renders the component', () => {

            expect(toJson(wrapper)).toMatchSnapshot();

        });
    });

    
    describe('onClick()',()=>{
        test('Delete recipe item',()=>{
            wrapper.find('WithStyles(IconButton)').props().onClick(event);
            expect(mockdeleteRecipefn).toHaveBeenCalled();
        });
    });

    describe('onEntring()', () => {
        test('set loader to true and calls to getRecipes()', () => {
            wrapper.find('Collapse').props().onEntering();
            expect(wrapper.state('loader')).toEqual(true);
            expect(mockgetRecipessfn).toHaveBeenCalled();
        });
    });

    
});