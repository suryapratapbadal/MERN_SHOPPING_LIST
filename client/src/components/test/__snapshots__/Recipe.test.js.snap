// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Recipe /> render() renders the component 1`] = `
<div>
  <Collapse
    appear={false}
    enter={true}
    exit={true}
    in={false}
    isOpen={true}
    mountOnEnter={false}
    onEnter={[Function]}
    onEntered={[Function]}
    onEntering={[Function]}
    onExit={[Function]}
    onExited={[Function]}
    onExiting={[Function]}
    style={
      Object {
        "marginTop": "1rem",
      }
    }
    tag="div"
    timeout={350}
    unmountOnExit={false}
  >
    <div
      className={Object {}}
    >
      <WithStyles(List)
        component="div"
        disablePadding={true}
        key="1"
      >
        <WithStyles(ListItem)
          button={true}
          className={Object {}}
        >
          <WithStyles(ListItemIcon)>
            <WithStyles(IconButton)
              aria-label="Delete"
              color="secondary"
              disableRipple={true}
              onClick={[Function]}
              style={
                Object {
                  "marginRight": "1rem",
                }
              }
            >
              <pure(DeleteIcon) />
            </WithStyles(IconButton)>
          </WithStyles(ListItemIcon)>
          <WithStyles(ListItemText)
            inset={true}
            primary="Tea"
          />
        </WithStyles(ListItem)>
      </WithStyles(List)>
    </div>
  </Collapse>
</div>
`;
